<a href="https://colab.research.google.com/github/shi-star/LSTM_IMDB_review/blob/main/LSTM_Imdb_review.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [19]:
from keras.datasets import imdb #Dataset of imdb
from keras.models import Sequential #Sequential the model 
from keras.layers import Dense  #Simple neural network layer
from keras.layers import LSTM   #sequence processing
from keras.layers.embeddings import Embedding  #to convert words into vectors fro processing
from keras.preprocessing import  sequence # for zero padding of small sentence so that it will be able to create batches of sequence of sentences of different length
import warnings
warnings.filterwarnings("ignore")

**Imported the required libraries and packages.**

In [3]:
top_words = 5000
(X_train, y_train),  (X_test, y_test) = imdb.load_data(num_words=top_words)

**We are taking only 5000 unique words from imdb dataset available in the keras.**

In [4]:
print(X_train.shape)
print(X_test.shape)

(25000,)
(25000,)


**Let's check the shape of the X_train, X_test**

In [5]:
print(X_train[0])

[1, 14, 22, 16, 43, 530, 973, 1622, 1385, 65, 458, 4468, 66, 3941, 4, 173, 36, 256, 5, 25, 100, 43, 838, 112, 50, 670, 2, 9, 35, 480, 284, 5, 150, 4, 172, 112, 167, 2, 336, 385, 39, 4, 172, 4536, 1111, 17, 546, 38, 13, 447, 4, 192, 50, 16, 6, 147, 2025, 19, 14, 22, 4, 1920, 4613, 469, 4, 22, 71, 87, 12, 16, 43, 530, 38, 76, 15, 13, 1247, 4, 22, 17, 515, 17, 12, 16, 626, 18, 2, 5, 62, 386, 12, 8, 316, 8, 106, 5, 4, 2223, 2, 16, 480, 66, 3785, 33, 4, 130, 12, 16, 38, 619, 5, 25, 124, 51, 36, 135, 48, 25, 1415, 33, 6, 22, 12, 215, 28, 77, 52, 5, 14, 407, 16, 82, 2, 8, 4, 107, 117, 2, 15, 256, 4, 2, 7, 3766, 5, 723, 36, 71, 43, 530, 476, 26, 400, 317, 46, 7, 4, 2, 1029, 13, 104, 88, 4, 381, 15, 297, 98, 32, 2071, 56, 26, 141, 6, 194, 2, 18, 4, 226, 22, 21, 134, 476, 26, 480, 5, 144, 30, 2, 18, 51, 36, 28, 224, 92, 25, 104, 4, 226, 65, 16, 38, 1334, 88, 12, 16, 283, 5, 16, 4472, 113, 103, 32, 15, 16, 2, 19, 178, 32]


**Here is the first sentence of our train data, however the words in the sentence has been presented as the index number of them.**

In [6]:
unique_word_dictionary = imdb.get_word_index()

**Got from imdb. to check the index number of each word.**

In [7]:
reverse_index_word_dictionary = {}
for i in unique_word_dictionary:
  if not unique_word_dictionary[i] in reverse_index_word_dictionary:
    reverse_index_word_dictionary[unique_word_dictionary[i]] = i

**Created the reverse dictionary so that we can check from index to words.**

In [ ]:
reverse_index_word_dictionary


We will see such type of dictionary, didn't show the output in notebook as it is such huge dictionary. And takes a lot of space.

12998: 'aragorn',<br>
 52272: 'hawker',<br>
 52273: 'hawkes',<br>
 3975: 'explosions',<br>
 8059: 'loren',<br>
 52274: "pyle's",<br>
 52274: "pyle's",<br>
 6704: 'shootout',<br>
 18517: "mike's",<br>
 52275: "driscoll's",<br>
 52275: "driscoll's",<br>
 40935: 'cogsworth',<br>
 52276: "britian's",<br>
 34744: 'childs',<br>
 52277: "portrait's",<br>
 3626: 'chain',<br>

**We can see the first sentence.**

In [9]:
for i in X_train[0]:
  print(reverse_index_word_dictionary[i])

the
as
you
with
out
themselves
powerful
lets
loves
their
becomes
reaching
had
journalist
of
lot
from
anyone
to
have
after
out
atmosphere
never
more
room
and
it
so
heart
shows
to
years
of
every
never
going
and
help
moments
or
of
every
chest
visual
movie
except
her
was
several
of
enough
more
with
is
now
current
film
as
you
of
mine
potentially
unfortunately
of
you
than
him
that
with
out
themselves
her
get
for
was
camp
of
you
movie
sometimes
movie
that
with
scary
but
and
to
story
wonderful
that
in
seeing
in
character
to
of
70s
and
with
heart
had
shadows
they
of
here
that
with
her
serious
to
have
does
when
from
why
what
have
critics
they
is
you
that
isn't
one
will
very
to
as
itself
with
other
and
in
of
seen
over
and
for
anyone
of
and
br
show's
to
whether
from
than
out
themselves
history
he
name
half
some
br
of
and
odd
was
two
most
of
mean
for
1
any
an
boat
she
he
should
is
thought
and
but
of
script
you
not
while
history
he
heart
to
real
at
and
but
when
from
one
bit
then
have
two
of
script
t

In [10]:
len(X_train[0])

218

**We can see the length of the first sentence is 218 words.**

In [13]:
max_review_length = 500
X_train = sequence.pad_sequences(X_train, maxlen=max_review_length)
X_test = sequence.pad_sequences(X_test, maxlen=max_review_length)

**Here we have taken only the sentence having max words 500 and less than 500. so all the sentences which have more than 500 words in them, will be truncated and having less than 500 words, will get zero padded.**

In [14]:
print(X_train[0])

[   0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0    0    0    0    0    0    0    0    0    0    0    0    0    0
    0 

**We can see the first sentence is of 218 length, so as it is less than 500 so zero padding has been applied here.**

In [28]:
embedding_vector_length = 32 # Embedding vector is used by the embedding layer internally
model = Sequential()
model.add(Embedding(top_words, embedding_vector_length, input_length=500))
model.add(LSTM(100))
model.add(Dense(1, activation='sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])
print(model.summary())

Model: "sequential_1"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
embedding_1 (Embedding)      (None, 500, 32)           160000    
_________________________________________________________________
lstm_1 (LSTM)                (None, 100)               53200     
_________________________________________________________________
dense_1 (Dense)              (None, 1)                 101       
Total params: 213,301
Trainable params: 213,301
Non-trainable params: 0
_________________________________________________________________
None


**we have built our model now.** 


In [29]:
model.fit(X_train, y_train, validation_data=(X_test, y_test), epochs=5, batch_size=64)

Epoch 1/5
391/391 [==============================] - 284s 680ms/step - loss: 0.5771 - accuracy: 0.6804 - val_loss: 0.3493 - val_accuracy: 0.8553
Epoch 2/5
391/391 [==============================] - 266s 681ms/step - loss: 0.3121 - accuracy: 0.8745 - val_loss: 0.3491 - val_accuracy: 0.8544
Epoch 3/5
391/391 [==============================] - 266s 682ms/step - loss: 0.2674 - accuracy: 0.8974 - val_loss: 0.3443 - val_accuracy: 0.8523
Epoch 4/5
391/391 [==============================] - 266s 679ms/step - loss: 0.2319 - accuracy: 0.9137 - val_loss: 0.3247 - val_accuracy: 0.8642
Epoch 5/5
391/391 [==============================] - 267s 684ms/step - loss: 0.2072 - accuracy: 0.9226 - val_loss: 0.3528 - val_accuracy: 0.8616


**We have fit the model.**

In [30]:
scores = model.evaluate(X_test, y_test, verbose=0)
print("Accuracy: %.2f%%" % (scores[1]*100))

Accuracy: 86.16%


**We can see the accuracy, We can try other things to increase the accuracy, like adding layers or increasing the neurons in the LSTM.** <br>
However at this point it is just for learning purpose , I am okay with 86% accuracy:)